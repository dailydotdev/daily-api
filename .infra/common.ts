export interface WorkerArgs {
  enableMessageOrdering?: boolean;
  ackDeadlineSeconds?: number;
  expirationPolicy?: {
    ttl: string;
  };
  deadLetterPolicy?: {
    deadLetterTopic: string;
    maxDeliveryAttempts: number;
  };
}

export interface Worker {
  topic: string;
  subscription: string;
  args?: WorkerArgs;
}

export const digestDeadLetter = 'api.v1.personalized-digest-email-dead-letter';

export const workers: Worker[] = [
  {
    topic: 'views',
    subscription: 'add-views-v2',
  },
  {
    topic: 'user-updated',
    subscription: 'api.user-updated-cio',
  },
  {
    topic: 'user-deleted',
    subscription: 'api.user-deleted-cio',
  },
  {
    topic: 'comment-upvoted',
    subscription: 'comment-upvoted-rep',
  },
  {
    topic: 'comment-upvote-canceled',
    subscription: 'comment-upvote-canceled-rep',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-scout-matched-slack-v2',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-freeform-images',
  },
  {
    topic: 'api.v1.post-content-edited',
    subscription: 'api.post-edited-freeform-images',
  },
  {
    topic: 'comment-commented',
    subscription: 'comment-commented-slack-message',
  },
  {
    topic: 'post-upvoted',
    subscription: 'post-upvoted-rep',
  },
  {
    topic: 'post-upvote-canceled',
    subscription: 'post-upvote-canceled-rep',
  },
  {
    topic: 'api.v1.post-downvoted',
    subscription: 'api.post-downvoted-rep',
  },
  {
    topic: 'api.v1.post-downvote-canceled',
    subscription: 'api.post-downvote-canceled-rep',
  },
  {
    topic: 'post-commented',
    subscription: 'api.post-commented-images',
  },
  {
    topic: 'post-commented',
    subscription: 'post-commented-redis',
  },
  {
    topic: 'post-commented',
    subscription: 'post-commented-slack-message',
  },
  {
    topic: 'post-commented',
    subscription: 'api.vordr-post-comment-prevented',
  },
  {
    topic: 'post-upvoted',
    subscription: 'post-upvoted-redis',
  },
  {
    topic: 'post-banned-or-removed',
    subscription: 'post-banned-rep',
  },
  {
    topic: 'post-banned-or-removed',
    subscription: 'api.post-deleted-comments-cleanup',
  },
  {
    topic: 'pub-request',
    subscription: 'pub-request-rep',
  },
  {
    topic: 'username-changed',
    subscription: 'username-changed-api',
  },
  {
    topic: 'update-comments',
    subscription: 'update-comments-mention',
  },
  {
    topic: 'api.changes',
    subscription: 'api-cdc',
    args: { enableMessageOrdering: true },
  },
  {
    topic: 'api.v1.cdc-notifications',
    subscription: 'api.cdc-notifications',
  },
  {
    topic: 'api.v1.new-notification',
    subscription: 'api.new-notification-real-time',
  },
  {
    topic: 'api.v1.new-notification',
    subscription: 'api.new-notification-mail',
  },
  {
    topic: 'api.v1.new-notification',
    subscription: 'api.new-notification-push',
  },
  {
    topic: 'api.v1.source-privacy-updated',
    subscription: 'api.source-privacy-updated',
  },
  {
    topic: 'api.v1.content-image-deleted',
    subscription: 'api.delete-cloudinary-image',
  },
  {
    topic: 'api.v1.comment-edited',
    subscription: 'api.comment-edited-images',
  },
  {
    topic: 'api.v1.comment-deleted',
    subscription: 'api.comment-deleted-images',
  },
  // Notifications
  {
    topic: 'community-link-rejected',
    subscription: 'api.community-picks-failed-notification',
  },
  {
    topic: 'community-link-access',
    subscription: 'api.community-picks-granted-notification',
  },
  {
    topic: 'post-commented',
    subscription: 'api.article-new-comment-notification.post-commented',
  },
  {
    topic: 'comment-commented',
    subscription: 'api.article-new-comment-notification.comment-commented',
  },
  {
    topic: 'api.v1.post-bookmark-reminder',
    subscription: 'api.post-bookmark-reminder-notification',
  },
  {
    topic: 'comment-commented',
    subscription: 'api.comment-commented-images',
  },
  {
    topic: 'post-upvoted',
    subscription: 'api.article-upvote-milestone-notification',
  },
  {
    topic: 'send-analytics-report',
    subscription: 'api.article-analytics-notification',
  },
  {
    topic: 'post-banned-or-removed',
    subscription: 'api.article-report-approved-notification',
  },
  {
    topic: 'pub-request',
    subscription: 'api.source-request-notification',
  },
  {
    topic: 'api.v1.new-post-mention',
    subscription: 'api.post-mention-notification',
  },
  {
    topic: 'api.v1.new-comment-mention',
    subscription: 'api.comment-mention-notification',
  },
  {
    topic: 'user-reputation-updated',
    subscription: 'api.user-reputation-updated-notification',
  },
  {
    topic: 'comment-commented',
    subscription: 'api.comment-reply-notification',
  },
  {
    topic: 'comment-upvoted',
    subscription: 'api.comment-upvote-milestone-notification',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-added-notification-v2',
  },
  {
    topic: 'api.v1.member-joined-source',
    subscription: 'api.member-joined-source-notification',
  },
  {
    topic: 'api.v1.user-created',
    subscription: 'api.user-created-cio',
  },
  {
    topic: 'api.v1.banner-added',
    subscription: 'api.banner-added',
  },
  {
    topic: 'api.v1.banner-deleted',
    subscription: 'api.banner-deleted',
  },
  {
    topic: 'yggdrasil.v1.content-published',
    subscription: 'api.content-published',
  },
  {
    topic: 'api.v1.source-member-role-changed',
    subscription: 'api.source-member-role-changed-notification',
  },
  {
    topic: 'analytics-api.v1.experiment-allocated',
    subscription: 'api.experiment-allocated',
  },
  {
    topic: 'api.v1.source-created',
    subscription: 'api.source-created-squad-user-action',
  },
  {
    topic: 'api.v1.post-collection-updated',
    subscription: 'api.post-collection-updated-notification',
  },
  {
    topic: 'api.v1.user-readme-updated',
    subscription: 'api.user-readme-images',
  },
  {
    topic: 'api.v1.user-created',
    subscription: 'api.user-created-personalized-digest-send-type',
  },
  {
    topic: 'api.v1.comment-downvoted',
    subscription: 'api.comment-downvoted-rep',
  },
  {
    topic: 'api.v1.comment-downvote-canceled',
    subscription: 'api.comment-downvote-canceled-rep',
  },
  {
    topic: 'api.v1.squad-public-request',
    subscription: 'api.v1.squad-public-request-notification',
  },
  {
    topic: 'api.v1.user-streak-updated',
    subscription: 'api.user-streak-updated-cio',
  },
  {
    topic: 'api.v1.user-streak-updated',
    subscription: 'api.user-streak-reset-notification',
  },
  {
    topic: 'api.v1.squad-featured-updated',
    subscription: 'api.squad-featured-updated-notification',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-added-slack-channel-send',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.vordr-post-prevented',
  },
  {
    topic: 'api.v1.user-company-approved',
    subscription: 'api.user-company-approved-cio',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-added-user-notification',
  },
  {
    topic: 'api.v1.user-top-reader',
    subscription: 'api.user-top-reader-added',
  }
];

export const personalizedDigestWorkers: Worker[] = [
  {
    topic: 'api.v1.generate-personalized-digest',
    subscription: 'api.personalized-digest-email',
  },
  {
    topic: digestDeadLetter,
    subscription: 'api.personalized-digest-email-dead-letter-log',
  },
];
