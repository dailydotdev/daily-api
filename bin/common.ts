import fastq from 'fastq';
import { ReadStream } from 'fs';

const QUEUE_CONCURRENCY = 1;

export const runInQueueStream = async <T>(
  stream: ReadStream,
  callback: (props: T) => Promise<void>,
  queue = QUEUE_CONCURRENCY,
) => {
  let insertCount = 0;
  const insertQueue = fastq.promise(async (props: T) => {
    await callback(props);

    insertCount += 1;
  }, queue);

  stream.on('data', insertQueue.push);

  await new Promise((resolve, reject) => {
    stream.on('error', reject);
    stream.on('end', resolve);
  });
  await insertQueue.drained();
  console.log('update finished with a total of: ', insertCount);
};
