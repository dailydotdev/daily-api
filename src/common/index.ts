export * from './base64';
export * from './cloudinary';
export * from './crypto';
export * from './feedGenerator';
export * from './healthCheck';
export * from './pagination';
export * from './pubsub';
export * from './reputation';
export * from './slack';
export * from './twitter';
export * from './users';
export * from './mailing';
export * from './post';
export * from './links';
export * from './utils';
export * from './typedPubsub';
export * from './personalizedDigest';
export * from './vote';
export * from './feed';
export * from './constants';
