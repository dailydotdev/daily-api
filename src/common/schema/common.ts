import { z } from 'zod';

/**
 * Common pagination arguments schema for cursor-based pagination
 * Use with .extend() to add to your query schemas
 */
export const paginationSchema = z.object({
  /**
   * Paginate after opaque cursor
   */
  after: z.string().optional(),
  /**
   * Number of items to return (first N items)
   */
  first: z.number().max(500).optional().default(100),
});

export type PaginationArgs = z.infer<typeof paginationSchema>;
