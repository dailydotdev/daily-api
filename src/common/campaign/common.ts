import { ValidationError } from 'apollo-server-errors';
import {
  Campaign,
  CampaignPost,
  CampaignType,
  type ConnectionManager,
} from '../../entity';
import { UserTransaction } from '../../entity/user/UserTransaction';
import { parseBigInt } from '../utils';
import { TransferError } from '../../errors';
import { transferCores, throwUserTransactionError, getBalance } from '../njord';
import type { AuthContext } from '../../Context';
import type { EntityManager } from 'typeorm';
import { CAMPAIGN_VALIDATION_SCHEMA } from '../schema/campaigns';
import { getSourceTags } from './source';
import { generatePostBoostEmail, getPostTags } from './post';
import type { NotificationBuilder } from '../../notifications/builder';
import { NotificationIcon } from '../../notifications/icons';
import { notificationsLink } from '../links';
import type { NotificationCampaignContext } from '../../notifications';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import { queryReadReplica } from '../queryReadReplica';

export interface StartCampaignArgs {
  value: string;
  duration: number;
  budget: number;
}

export interface StartCampaignMutationArgs {
  ctx: AuthContext;
  args: StartCampaignArgs;
}

export const validateCampaignArgs = (
  args: Pick<StartCampaignArgs, 'budget' | 'duration'>,
) => {
  const result = CAMPAIGN_VALIDATION_SCHEMA.safeParse(args);

  if (result.error) {
    throw new ValidationError(result.error.issues[0].message);
  }
};

interface StartCampaignTransferCoresProps {
  ctx: AuthContext;
  campaignId: string;
  userTransaction: UserTransaction;
  manager: EntityManager;
}

export const startCampaignTransferCores = async ({
  ctx,
  campaignId,
  userTransaction,
  manager,
}: StartCampaignTransferCoresProps) => {
  if (ctx.isTeamMember) {
    return {
      transaction: {
        referenceId: campaignId,
        transactionId: userTransaction.id,
        balance: { amount: (await getBalance({ userId: ctx.userId })).amount },
      },
    };
  }

  try {
    const transfer = await transferCores({
      ctx,
      transaction: userTransaction,
      entityManager: manager,
    });

    return {
      transfer,
      transaction: {
        referenceId: campaignId,
        transactionId: userTransaction.id,
        balance: {
          amount: parseBigInt(transfer.senderBalance?.newBalance),
        },
      },
    };
  } catch (error) {
    if (error instanceof TransferError) {
      await throwUserTransactionError({
        ctx,
        entityManager: manager,
        error,
        transaction: userTransaction,
      });
    }

    throw error;
  }
};

export const stopCampaignTransferCores = async ({
  ctx,
  campaignId,
  userTransaction,
  manager,
}: StartCampaignTransferCoresProps) => {
  if (ctx.isTeamMember) {
    return {
      transaction: {
        referenceId: campaignId,
        transactionId: userTransaction.id,
        balance: { amount: (await getBalance({ userId: ctx.userId })).amount },
      },
    };
  }

  try {
    const transfer = await transferCores({
      ctx,
      transaction: userTransaction,
      entityManager: manager,
    });

    return {
      transfer,
      transaction: {
        referenceId: campaignId,
        transactionId: userTransaction.id,
        balance: {
          amount: parseBigInt(transfer.receiverBalance?.newBalance),
        },
      },
    };
  } catch (error) {
    if (error instanceof TransferError) {
      await throwUserTransactionError({
        ctx,
        entityManager: manager,
        error,
        transaction: userTransaction,
      });
    }

    throw error;
  }
};

export interface StopCampaignProps {
  campaign: Campaign;
  ctx: AuthContext;
}

export const getReferenceTags = (
  con: ConnectionManager,
  type: CampaignType,
  referenceId: string,
) => {
  switch (type) {
    case CampaignType.Post:
      return getPostTags(con, referenceId);
    case CampaignType.Squad:
      return getSourceTags(con, referenceId);
    default:
      throw new ValidationError('Unknown campaign type to estimate reach');
  }
};

export enum CampaignUpdateEvent {
  Started = 'CAMPAIGN_STARTED',
  Completed = 'CAMPAIGN_COMPLETED',
  StatsUpdated = 'STATS_UPDATED',
  BudgetUpdated = 'BUDGET_UPDATED',
}

export interface CampaignCompleted {
  budget: string;
}

export interface CampaignStatsUpdate {
  impressions: number;
  clicks: number;
  unique_users: number;
}

export interface CampaignStateUpdate {
  budget: string; // used budget
}

export interface CampaignUpdateEventArgs {
  campaignId: string;
  event: CampaignUpdateEvent;
  unique_users: number;
  data: CampaignCompleted | CampaignStatsUpdate | CampaignStateUpdate;
  d_update: number;
}

export const generateCampaignCompletedNotification = (
  builder: NotificationBuilder,
  ctx: NotificationCampaignContext,
) => {
  const { campaign, source, event, user } = ctx;

  const nb = builder
    .icon(NotificationIcon.DailyDev)
    .referenceCampaign(ctx)
    .targetUrl(notificationsLink)
    .setTargetUrlParameter([['c_id', campaign.id]])
    .uniqueKey(`${campaign.id}-${user.id}-${event}`);

  switch (campaign.type) {
    case CampaignType.Post:
      return nb.avatarUser(user);
    case CampaignType.Squad:
      if (!source) {
        throw new Error(
          `Can't generate Squad Campaign Notification without the Squad`,
        );
      }
      return nb.avatarSource(source);
    default:
      throw new Error(
        `Unable to generate notification for unknown type: ${campaign.type}`,
      );
  }
};

export const generateCampaignCompletedEmail: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await con
    .getRepository(Campaign)
    .findOneBy({ id: notification.referenceId });

  if (!campaign) {
    return null;
  }

  switch (campaign.type) {
    case CampaignType.Post:
      return generatePostBoostEmail({
        con,
        postId: campaign.referenceId,
        notification,
        campaign,
      });
    // TODO: MI-1007 - generate email once template id is provided
    default:
      return null;
  }
};

export interface UserCampaignStats {
  impressions: number;
  clicks: number;
  spend: number;
  users: number;
}

export const getUserCampaignStats = async (
  ctx: AuthContext,
): Promise<UserCampaignStats> => {
  const result = await queryReadReplica(ctx.con, ({ queryRunner }) =>
    queryRunner.manager
      .getRepository(CampaignPost)
      .createQueryBuilder('c')
      .select(`SUM(COALESCE((c.flags->>'impressions')::int, 0))`, 'impressions')
      .addSelect(`SUM(COALESCE((c.flags->>'users')::int, 0))`, 'users')
      .addSelect(`SUM(COALESCE((c.flags->>'clicks')::int, 0))`, 'clicks')
      .addSelect(`SUM(COALESCE((c.flags->>'spend')::int, 0))`, 'spend')
      .where(`c."userId" = :user`, { user: ctx.userId })
      .getRawOne(),
  );

  return {
    clicks: result.clicks ?? 0,
    impressions: result.impressions ?? 0,
    users: result.users ?? 0,
    spend: result.spend ?? 0,
  };
};
