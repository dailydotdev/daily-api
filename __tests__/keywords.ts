import {
  disposeGraphQLTesting,
  GraphQLTestClient,
  GraphQLTestingState,
  initializeGraphQLTesting,
  MockContext,
  Mutation,
  Query,
  saveFixtures,
  testMutationErrorCode,
  testQueryErrorCode,
} from './helpers';
import { Roles } from '../src/roles';
import { ArticlePost, Keyword, PostKeyword, Source } from '../src/entity';
import { sourcesFixture } from './fixture/source';
import { postsFixture } from './fixture/post';
import { DataSource } from 'typeorm';
import createOrGetConnection from '../src/db';
import { updateFlagsStatement } from '../src/common';

let con: DataSource;
let state: GraphQLTestingState;
let client: GraphQLTestClient;
let loggedUser: string = null;
let roles: Roles[] = [];

beforeAll(async () => {
  con = await createOrGetConnection();
  state = await initializeGraphQLTesting(
    () => new MockContext(con, loggedUser, false, roles),
  );
  client = state.client;
});

beforeEach(async () => {
  jest.resetAllMocks();
  loggedUser = null;
  roles = [];

  await saveFixtures(con, Source, sourcesFixture);
  await saveFixtures(con, ArticlePost, postsFixture);
});

afterAll(() => disposeGraphQLTesting(state));

const testModeratorQueryAuthorization = (query: Query): Promise<void> => {
  roles = [];
  loggedUser = '1';
  return testQueryErrorCode(client, query, 'FORBIDDEN');
};

const testModeratorMutationAuthorization = (
  mutation: Mutation,
): Promise<void> => {
  roles = [];
  loggedUser = '1';
  return testMutationErrorCode(client, mutation, 'FORBIDDEN');
};

describe('query randomPendingKeyword', () => {
  const QUERY = `{
    randomPendingKeyword {
      value, status, occurrences
    }
  }`;

  it('should not authorize when not moderator', () =>
    testModeratorQueryAuthorization({
      query: QUERY,
    }));

  it('should return an eligible pending keyword', async () => {
    roles = [Roles.Moderator];
    loggedUser = '1';
    await con
      .getRepository(Keyword)
      .save([
        { value: 'nodejs', status: 'allow', occurrences: 200 },
        { value: 'react' },
        { value: 'go', occurrences: 100 },
      ]);
    const res = await client.query(QUERY);
    expect(res.errors).toBeFalsy();
    expect(res.data).toMatchSnapshot();
  });
});

describe('query countPendingKeywords', () => {
  const QUERY = `{ countPendingKeywords }`;

  it('should not authorize when not moderator', () =>
    testModeratorQueryAuthorization({
      query: QUERY,
    }));

  it('should return the number of pending keywords', async () => {
    roles = [Roles.Moderator];
    loggedUser = '1';
    await con
      .getRepository(Keyword)
      .save([
        { value: 'nodejs', status: 'allow', occurrences: 200 },
        { value: 'react', occurrences: 300 },
        { value: 'go', occurrences: 100 },
        { value: 'vuejs' },
      ]);
    const res = await client.query(QUERY);
    expect(res.errors).toBeFalsy();
    expect(res.data.countPendingKeywords).toEqual(2);
  });
});

describe('query searchKeywords', () => {
  const QUERY = `
  query SearchKeywords($query: String!) {
    searchKeywords(query: $query) {
      query, hits { value, status, occurrences }
    }
  }`;

  it('should not authorize when not moderator', () =>
    testModeratorQueryAuthorization({
      query: QUERY,
      variables: { query: 'script' },
    }));

  it('should return search results', async () => {
    roles = [Roles.Moderator];
    loggedUser = '1';
    await con.getRepository(Keyword).save([
      { value: 'javascript', status: 'allow', occurrences: 20 },
      { value: 'java', status: 'allow' },
      { value: 'typescript', occurrences: 50 },
      { value: 'nativescript', status: 'allow', occurrences: 80 },
    ]);
    const res = await client.query(QUERY, { variables: { query: 'script' } });
    expect(res.errors).toBeFalsy();
    expect(res.data).toMatchSnapshot();
  });
});

describe('query keyword', () => {
  const QUERY = `
  query Keyword($value: String!) {
    keyword(value: $value) {
      value, status, occurrences
    }
  }`;

  it('should return keyword', async () => {
    roles = [Roles.Moderator];
    loggedUser = '1';
    await con.getRepository(Keyword).save([
      { value: 'nodejs', status: 'allow', occurrences: 200 },
      { value: 'react', occurrences: 300 },
    ]);
    const res = await client.query(QUERY, { variables: { value: 'nodejs' } });
    expect(res.errors).toBeFalsy();
    expect(res.data).toMatchSnapshot();
  });

  it('should return null when keyword does not exist', async () => {
    roles = [Roles.Moderator];
    loggedUser = '1';
    await con.getRepository(Keyword).save([
      { value: 'nodejs', status: 'allow', occurrences: 200 },
      { value: 'react', occurrences: 300 },
    ]);
    const res = await client.query(QUERY, { variables: { value: 'go' } });
    expect(res.errors).toBeFalsy();
    expect(res.data).toMatchSnapshot();
  });
});

describe('mutation allowKeyword', () => {
  const MUTATION = `
  mutation AllowKeyword($keyword: String!) {
    allowKeyword(keyword: $keyword) {
      _
    }
  }`;

  it('should not authorize when not moderator', () =>
    testModeratorMutationAuthorization({
      mutation: MUTATION,
      variables: { keyword: 'java' },
    }));

  it('should allow existing keyword', async () => {
    roles = [Roles.Moderator];
    loggedUser = '1';
    await con.getRepository(Keyword).save([{ value: 'java', occurrences: 20 }]);
    const res = await client.mutate(MUTATION, {
      variables: { keyword: 'java' },
    });
    expect(res.errors).toBeFalsy();
    const keywords = await con.getRepository(Keyword).find({
      select: ['value', 'status', 'occurrences'],
      order: { value: 'ASC' },
    });
    expect(keywords).toMatchSnapshot();
  });

  it('should create a new keyword and allow it', async () => {
    roles = [Roles.Moderator];
    loggedUser = '1';
    const res = await client.mutate(MUTATION, {
      variables: { keyword: 'java' },
    });
    expect(res.errors).toBeFalsy();
    const keywords = await con.getRepository(Keyword).find({
      select: ['value', 'status', 'occurrences'],
      order: { value: 'ASC' },
    });
    expect(keywords).toMatchSnapshot();
  });
});

describe('mutation denyKeyword', () => {
  const MUTATION = `
  mutation DenyKeyword($keyword: String!) {
    denyKeyword(keyword: $keyword) {
      _
    }
  }`;

  it('should not authorize when not moderator', () =>
    testModeratorMutationAuthorization({
      mutation: MUTATION,
      variables: { keyword: 'java' },
    }));

  it('should deny existing keyword', async () => {
    roles = [Roles.Moderator];
    loggedUser = '1';
    await con.getRepository(Keyword).save([{ value: 'java', occurrences: 20 }]);
    const res = await client.mutate(MUTATION, {
      variables: { keyword: 'java' },
    });
    expect(res.errors).toBeFalsy();
    const keywords = await con.getRepository(Keyword).find({
      select: ['value', 'status', 'occurrences'],
      order: { value: 'ASC' },
    });
    expect(keywords).toMatchSnapshot();
  });

  it('should create a new keyword and deny it', async () => {
    roles = [Roles.Moderator];
    loggedUser = '1';
    const res = await client.mutate(MUTATION, {
      variables: { keyword: 'java' },
    });
    expect(res.errors).toBeFalsy();
    const keywords = await con.getRepository(Keyword).find({
      select: ['value', 'status', 'occurrences'],
      order: { value: 'ASC' },
    });
    expect(keywords).toMatchSnapshot();
  });
});

describe('mutation setKeywordAsSynonym', () => {
  const MUTATION = `
  mutation SetKeywordAsSynonym($keywordToUpdate: String!, $originalKeyword: String!) {
    setKeywordAsSynonym(keywordToUpdate: $keywordToUpdate, originalKeyword: $originalKeyword) {
      _
    }
  }`;

  it('should not authorize when not moderator', () =>
    testModeratorMutationAuthorization({
      mutation: MUTATION,
      variables: { keywordToUpdate: 'react', originalKeyword: 'reactjs' },
    }));

  it('should set keyword as synonym and rename existing occurrences', async () => {
    roles = [Roles.Moderator];
    loggedUser = '1';
    await con.getRepository(Keyword).save([
      { value: 'reactjs', occurrences: 20, status: 'allow' },
      { value: 'react', occurrences: 100 },
    ]);
    await con.getRepository(PostKeyword).save([
      { postId: 'p1', keyword: 'react' },
      { postId: 'p1', keyword: 'javascript' },
      { postId: 'p2', keyword: 'react' },
      { postId: 'p2', keyword: 'typescript' },
      { postId: 'p3', keyword: 'reactjs' },
    ]);
    const res = await client.mutate(MUTATION, {
      variables: { keywordToUpdate: 'react', originalKeyword: 'reactjs' },
    });
    expect(res.errors).toBeFalsy();
    const keywords = await con.getRepository(Keyword).find({
      select: ['value', 'status', 'occurrences'],
      order: { value: 'ASC' },
    });
    const postKeywords = await con.getRepository(PostKeyword).find({
      order: {
        postId: 'ASC',
        keyword: 'ASC',
      },
    });
    expect(keywords).toMatchSnapshot();
    expect(postKeywords).toMatchSnapshot();
  });

  it('should create keywords if they do not exist', async () => {
    roles = [Roles.Moderator];
    loggedUser = '1';
    await con.getRepository(PostKeyword).save([
      { postId: 'p1', keyword: 'react' },
      { postId: 'p1', keyword: 'javascript' },
      { postId: 'p2', keyword: 'react' },
      { postId: 'p2', keyword: 'typescript' },
      { postId: 'p3', keyword: 'reactjs' },
    ]);
    const res = await client.mutate(MUTATION, {
      variables: { keywordToUpdate: 'react', originalKeyword: 'reactjs' },
    });
    expect(res.errors).toBeFalsy();
    const keywords = await con.getRepository(Keyword).find({
      select: ['value', 'status', 'occurrences'],
      order: { value: 'ASC' },
    });
    const postKeywords = await con.getRepository(PostKeyword).find({
      order: {
        postId: 'ASC',
        keyword: 'ASC',
      },
    });
    expect(keywords).toMatchSnapshot();
    expect(postKeywords).toMatchSnapshot();
  });

  it('should ignore duplicates keywords when renaming', async () => {
    roles = [Roles.Moderator];
    loggedUser = '1';
    await con.getRepository(PostKeyword).save([
      { postId: 'p1', keyword: 'react' },
      { postId: 'p1', keyword: 'javascript' },
      { postId: 'p1', keyword: 'reactjs' },
      { postId: 'p2', keyword: 'react' },
      { postId: 'p2', keyword: 'typescript' },
      { postId: 'p3', keyword: 'reactjs' },
    ]);
    const res = await client.mutate(MUTATION, {
      variables: { keywordToUpdate: 'react', originalKeyword: 'reactjs' },
    });
    expect(res.errors).toBeFalsy();
    const postKeywords = await con.getRepository(PostKeyword).find({
      order: {
        postId: 'ASC',
        keyword: 'ASC',
      },
    });
    expect(postKeywords).toMatchSnapshot();
  });
});

describe('flags field', () => {
  const QUERY = `{
    keyword(value: "react") {
      flags {
        title
        description
      }
    }
  }`;

  beforeEach(async () => {
    await con.getRepository(Keyword).save([
      { value: 'react', status: 'allow', occurrences: 200 },
      { value: 'nodejs', status: 'allow', occurrences: 300 },
      { value: 'go', status: 'allow', occurrences: 100 },
    ]);
  });

  it('should return all the public flags for anonymous user', async () => {
    await con.getRepository(Keyword).update(
      { value: 'react' },
      {
        flags: updateFlagsStatement({
          title: 'React',
          description: 'React is a JS library',
        }),
      },
    );
    const res = await client.query(QUERY);
    expect(res.errors).toBeFalsy();
    expect(res.data.keyword.flags).toEqual({
      title: 'React',
      description: 'React is a JS library',
    });
  });

  it('should return null values for unset flags', async () => {
    const res = await client.query(QUERY);
    expect(res.data.keyword.flags).toEqual({
      title: null,
      description: null,
    });
  });

  it('should contain all default values in db query', async () => {
    const keyword = await con
      .getRepository(Keyword)
      .findOneBy({ value: 'react' });
    expect(keyword?.flags).toEqual({});
  });
});
