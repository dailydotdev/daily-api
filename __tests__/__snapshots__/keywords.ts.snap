// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutation allowKeyword should allow existing keyword 1`] = `
Array [
  Keyword {
    "flags": Object {},
    "occurrences": 20,
    "status": "allow",
    "value": "java",
  },
]
`;

exports[`mutation allowKeyword should create a new keyword and allow it 1`] = `
Array [
  Keyword {
    "flags": Object {},
    "occurrences": 1,
    "status": "allow",
    "value": "java",
  },
]
`;

exports[`mutation denyKeyword should create a new keyword and deny it 1`] = `
Array [
  Keyword {
    "flags": Object {},
    "occurrences": 1,
    "status": "deny",
    "value": "java",
  },
]
`;

exports[`mutation denyKeyword should deny existing keyword 1`] = `
Array [
  Keyword {
    "flags": Object {},
    "occurrences": 20,
    "status": "deny",
    "value": "java",
  },
]
`;

exports[`mutation setKeywordAsSynonym should create keywords if they do not exist 1`] = `
Array [
  Keyword {
    "flags": Object {},
    "occurrences": 1,
    "status": "synonym",
    "value": "react",
  },
  Keyword {
    "flags": Object {},
    "occurrences": 1,
    "status": "allow",
    "value": "reactjs",
  },
]
`;

exports[`mutation setKeywordAsSynonym should create keywords if they do not exist 2`] = `
Array [
  PostKeyword {
    "keyword": "javascript",
    "postId": "p1",
    "status": null,
  },
  PostKeyword {
    "keyword": "reactjs",
    "postId": "p1",
    "status": "allow",
  },
  PostKeyword {
    "keyword": "reactjs",
    "postId": "p2",
    "status": "allow",
  },
  PostKeyword {
    "keyword": "typescript",
    "postId": "p2",
    "status": null,
  },
  PostKeyword {
    "keyword": "reactjs",
    "postId": "p3",
    "status": "allow",
  },
]
`;

exports[`mutation setKeywordAsSynonym should ignore duplicates keywords when renaming 1`] = `
Array [
  PostKeyword {
    "keyword": "javascript",
    "postId": "p1",
    "status": null,
  },
  PostKeyword {
    "keyword": "reactjs",
    "postId": "p1",
    "status": "allow",
  },
  PostKeyword {
    "keyword": "reactjs",
    "postId": "p2",
    "status": "allow",
  },
  PostKeyword {
    "keyword": "typescript",
    "postId": "p2",
    "status": null,
  },
  PostKeyword {
    "keyword": "reactjs",
    "postId": "p3",
    "status": "allow",
  },
]
`;

exports[`mutation setKeywordAsSynonym should set keyword as synonym and rename existing occurrences 1`] = `
Array [
  Keyword {
    "flags": Object {},
    "occurrences": 100,
    "status": "synonym",
    "value": "react",
  },
  Keyword {
    "flags": Object {},
    "occurrences": 20,
    "status": "allow",
    "value": "reactjs",
  },
]
`;

exports[`mutation setKeywordAsSynonym should set keyword as synonym and rename existing occurrences 2`] = `
Array [
  PostKeyword {
    "keyword": "javascript",
    "postId": "p1",
    "status": null,
  },
  PostKeyword {
    "keyword": "reactjs",
    "postId": "p1",
    "status": "allow",
  },
  PostKeyword {
    "keyword": "reactjs",
    "postId": "p2",
    "status": "allow",
  },
  PostKeyword {
    "keyword": "typescript",
    "postId": "p2",
    "status": null,
  },
  PostKeyword {
    "keyword": "reactjs",
    "postId": "p3",
    "status": "allow",
  },
]
`;

exports[`query keyword should return keyword 1`] = `
Object {
  "keyword": Object {
    "occurrences": 200,
    "status": "allow",
    "value": "nodejs",
  },
}
`;

exports[`query keyword should return null when keyword does not exist 1`] = `
Object {
  "keyword": null,
}
`;

exports[`query randomPendingKeyword should return an eligible pending keyword 1`] = `
Object {
  "randomPendingKeyword": Object {
    "occurrences": 100,
    "status": "pending",
    "value": "go",
  },
}
`;

exports[`query searchKeywords should return search results 1`] = `
Object {
  "searchKeywords": Object {
    "hits": Array [
      Object {
        "occurrences": 80,
        "status": "allow",
        "value": "nativescript",
      },
      Object {
        "occurrences": 20,
        "status": "allow",
        "value": "javascript",
      },
    ],
    "query": "script",
  },
}
`;
